torch/csrc/jit/import.cpp

```diff
@@ -155,27 +161,162 @@ void buildBlock(const onnx_torch::GraphProto& graph_proto, Block* block,
   }
 }
 
-std::shared_ptr<Graph> buildGraph(const onnx_torch::GraphProto& graph_proto, std::vector<at::Tensor>& initializers) {
+class GraphDecoder : JitDecoder {
+ public:
+  std::shared_ptr<Graph> decode(const std::string& serialized_graph,
+                                std::vector<at::Tensor>& initializers);
+};
 
-  auto graph = buildGraph(graph_proto);
+std::shared_ptr<Graph> GraphDecoder::decode(
+    const std::string& serialized_graph,
+    std::vector<at::Tensor>& initializers) {
+  auto model_proto = onnx_torch::ModelProto();
+  model_proto.ParseFromString(serialized_graph);
 
-  for (auto tensor_ : graph_proto.initializer()) {
+  auto graph_proto = model_proto.graph();
+  auto graph = buildGraph(graph_proto);
+  for (auto &tensor_ : graph_proto.initializer()) {
     initializers.push_back(buildTensor(tensor_));
   }
-
   return graph;
 }
 
+class ModuleDecoder : JitDecoder {
+ public:
+  std::shared_ptr<script::Module> decode(
+      std::shared_ptr<script::Module> root_module,
+      const std::string& serialized_module,
+      const std::unordered_map<std::string, std::string>& storage_map);
+
+ private:
+  at::Tensor buildTensor(const onnx_torch::TensorProto& tensor_proto);
+
+  at::Tensor buildParameter(const onnx_torch::TensorProto& tensor_proto);
+
+  at::Storage* getStorage(at::ScalarType type, std::string name);
+
+  std::pair<std::shared_ptr<script::Module>, std::string> parseFullName(
+      std::shared_ptr<script::Module> root_module,
+      const std::string fullname);
+
+  const std::unordered_map<std::string, std::string> *storage_export_map_;
+  std::unordered_map<std::string, std::unique_ptr<at::Storage>> storage_map_;
+};
+
+at::Tensor ModuleDecoder::buildParameter(const onnx_torch::TensorProto& tensor_proto) {
```

Comment:
similar to encoder, we should not duplicate the logic between buildTensor and buildParameter.